streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamAggregate.java

```diff
@@ -53,13 +53,17 @@ public void enableSendingOldValues() {
     private class KStreamAggregateProcessor extends AbstractProcessor<K, V> {
 
         private KeyValueStore<K, T> store;
+        private boolean cached = false;
 
         @SuppressWarnings("unchecked")
         @Override
         public void init(ProcessorContext context) {
             super.init(context);
             store = (KeyValueStore<K, T>) context.getStateStore(storeName);
-            ((CachedStateStore) store).setFlushListener(new ForwardingCacheFlushListener<K, V>(context, sendOldValues));
+            if (store instanceof CachedStateStore) {
```

Comment:
i was hoping we wouldn't have this block of code and the subsequent block in `process` essentially duplicated though all the processors. Can we encapsulate the logic such that it is only done in a single place? i.e, We could have a class sth like:
```
class Forwarder<K, V> {
   public Forwarder(final StateStore store, final ProcessorContext context, final ForwardingCacheFlushListener forwardingListener) {
   this.cached = store instanceof CachedStateStore;
   if (this.cached) {
      ((CachedStateStore) store).setFlushListener(listener);
   }
   ...
}

 void maybeForward(final K key, final V newValue, final V oldValue) {
    if(!cached) {
        forwardingListener.apply(key, newValue, oldValue);
    }
 } 
}
```
Then we could construct this class in `init` and then in `process` we call `forwarder.maybeForward(key, newValue, oldValue)`

I think then we should be able to remove the duplication?